cmd/authgear/pgsearch/reindexer.go

```diff
@@ -0,0 +1,89 @@
+package pgsearch
+
+import (
+	"context"
+	"log"
+
+	"github.com/authgear/authgear-server/pkg/api/model"
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/lib/infra/db/appdb"
+	"github.com/authgear/authgear-server/pkg/lib/search/pgsearch"
+	"github.com/authgear/authgear-server/pkg/lib/search/reindex"
+)
+
+type Reindexer struct {
+	Handle *appdb.Handle
+	AppID  config.AppID
+
+	SourceProvider *reindex.SourceProvider
+}
+
+func (q *Reindexer) Reindex(ctx context.Context, store *pgsearch.Store) (err error) {
+
+	allUserIDs, err := q.reindex(ctx, store)
+	if err != nil {
+		return
+	}
+
+	deletedCount, err := q.cleanupDeletedUsers(ctx, store, allUserIDs)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("App (%v): %v indexed; %v deleted;\n", q.AppID, len(allUserIDs), deletedCount)
+	return nil
+}
+
+func (q *Reindexer) reindex(ctx context.Context, store *pgsearch.Store) (allUserIDs map[string]struct{}, err error) {
+	allUserIDs = make(map[string]struct{})
+
+	var first uint64 = 500
+	var after model.PageCursor = ""
+	var items []reindex.ReindexItem
+
+	for {
+		err = q.Handle.WithTx(ctx, func(ctx context.Context) (err error) {
```

Comment:
I suggest we do not open tx at the same time. Open a tx in appdb to fetch the source, close this tx. And then open a tx in searchdb to upsert.